### DISCUSSION TITLE
Future of XQuartz on macOS

### DISCUSSION THREAD

**Discussion by d235j** [Homebrew/discussions#205] [Everyday usage]
It appears that upstream has not been maintaining the XQuartz binary package, and is instead encouraging people to use the MacPorts `xorg-server` or `xorg-server-devel` port. [Version 2.7.11 was released in October 2016](https://www.xquartz.org/releases/index.html) — over four years ago.

[From their maintainer on the mailing list in March 2020](https://lists.macosforge.org/pipermail/xquartz-dev/2020-March/004072.html):
>> The last version of XQuartz available for download is from 2016.
>
> Yeah, that's the last time I had any free time to package everything up into a release, and nobody has stepped up to take that over.  If you want the latest versions, I suggest using MacPorts to install it.

[The official release information page](https://www.xquartz.org/releases/index.html) also points people to install it via MacPorts, stating that "This is currently tested and supported on Tiger and newer versions of OS X."

There are newer mailing list posts of people trying to build XQuartz with newer SDK, without much luck and with no one helping.

Based on this, I don't believe XQuartz.app will be updated for Apple Silicon anytime soon, unless someone else takes over.

[The MacPorts version of xorg-server](https://github.com/macports/macports-ports/blob/master/x11/xorg-server/Portfile) is maintained, currently at version 1.20.10. It does utilize several patches.

With formulas moving to no longer depend on XQuartz for the benefit of LinuxBrew et al, perhaps it would make sense to transition to installing Xorg on Mac via Homebrew, perhaps using the MacPorts Portfiles as a template? I think this needs further discussion.

**Comment by jonchang** [comment]
Yes, this would be nice to have, but to my knowledge no maintainers are currently working on this. You'll need to build new formulae for a few things (applewm and quartzwm at the minimum) and then create a formula for xorg-server and xinit. See https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request for instructions on how to get started.

**Reply by d235j** [reply]
I definitely can start working on it! I'm familiar with Git and pull requests; I just wanted to know if there have been any discussions among leadership whether this is the direction Homebrew wants to take or if it should be in the roadmap, as I don't want to spend significant time if it will be rejected.

It is also worth considering that it doesn't seem like XQuartz will have an Apple Silicon build anytime soon — which means it would be a blocker for Apple Silicon support for any formulas that require Xorg. I'm not yet certain how this will work prefixes, either — XQuartz installs itself in /opt/X11, but I'm not certain if the MacPorts-installed version does so, or what exactly happens on Apple Silicon. Once I have an Apple Silicon computer I'll be doing some testing.

Overall, I am disappointed that XQuartz upstream has not issued clearer guidance of the future of the XQuartz.app package; installing it from MacPorts is fine, but not everyone wants to use MacPorts.

**Comment by kencu** [comment]
xorg-server is building on arm64 <http://packages.macports.org/xorg-server/xorg-server-1.20.9_0.darwin_20.arm64.tbz2>.

I haven't installed it myself or tried it, but it is sitting right there.

If you don't want to install MacPorts as your permanent package manager, you can use MacPorts to build a new installer.

MacPorts can build "packages" for all the software a given port requires. At least a few of the current homebrew "cask" installs are built that way -- the wine ones, for example. 

Installing MacPorts into a custom prefix like ```/opt/xquartznew``` or whatever you like, then installing ```xorg-server``` into that prefix, and then running the packaging command will generate a standalone installer that will install xorg-server into that prefix (that is where XQuartz.app comes from, years ago).

Then there is just the matter of the install scripts that launch it automatically from the terminal.

I have been pondering helping Jeremy out by updating XQuartz.app for him, but as xorg-server works so well as it is, and is updated frequently on MacPorts, I just haven't bothered.

**Comment by kencu** [comment]
Oh, and I can personally verify that xorg-server works perfectly on 10.4 Tiger PPC to Big Sur. I've done some of the (minor) 10.4 PPC patches myself, but it's written in a very forgiving way, so very little is needed.

**Comment by d235j** [comment]
@kencu This is great, but some of us don't want to run MacPorts, and to be honest, some of us don't want to run Homebrew or a package manager either, but still need X! One major use is for X forwarding, but compiling software that requires X manually / without a package manager is another use.

While running XQuartz.app 2.7.11 in Rosetta 2 is an option for now for the X-forwarding use case, I don't consider it a good long-term solution, and this does not help for other use cases.

Therefore, it would be nice to obtain official guidance as to whether there will be an updated XQuartz.app or if instead people are expected to use MacPorts — in which case we'd want to add xorg-server to Homebrew, so that people don't need to run separate MacPorts and Homebrew installations on the same computer to natively use software that requires X.

**Comment by kencu** [comment]
I guess what I was trying to say (but it didn't come out clearly) is that if you don't want to install MacPorts, that someone could spend about an hour and make up a new, current XQuartz.app installer with MacPorts that would install XQuartz.app into ```/Applications``` and with all dependencies into ```/opt/anything```.

Jeremy, though, who did the last one, is just too busy. But it's really simple to do it.

**Reply by d235j** [reply]
That's fair, but in that case the issue would be getting the new, current XQuartz.app installer officially "blessed" and uploaded to xquartz.org, and added to the Sparkle updater feed, probably after a beta-test period.
The building and packaging could even be automated using Github Actions or something similar, but I'm not sure who could do the latter part other than Jeremy.

**Comment by kencu** [comment]
The building and packaging are done with MacPorts, so you don't need Github actions or anything else.

Any project, with all deps, can be built into a standalone installer with MacPorts.

**Reply by d235j** [reply]
@kencu 
GitHub Actions would be a way to automate the process on a CI infrastructure, rather than having to do it on one's local computer. It would still be using MacPorts for building and packaging into a standalone installer.

Again, this doesn't solve the issue of officially "blessing" and uploading the package to xquartz.org, which is what we need guidance on. We are not interested in distributing private/divergent builds of the .app. If such an official build is unlikely to happen, and users are told to use MacPorts instead, it would probably make the most sense for us to add `xorg-server` (and dependencies) to Homebrew, compiled in a similar way as the MacPorts `xorg-server` port. [Considering that the groundwork for compiling `xorg-server` is being done for Linux](https://github.com/Linuxbrew/homebrew-xorg/), this would be an option.

However, if there were an official updated build of XQuartz.app, we would prefer to use that.

**Comment by kencu** [comment]
The current XQuartz.app appears to be built with MacPorts supports, then packaged and uploaded to xquartz.org.

But there are no doubt other ways that it could be done.

**Comment by tlk** [comment]
> Based on this, I don't believe XQuartz.app will be updated for Apple Silicon

Check out https://github.com/XQuartz/XQuartz - it looks like a new beta version with Apple M1 support is about to be released.

**Comment by kencu** [comment]
There was extensive work on Xquartz after these posts, and it is now appearing 100% fully up-to-date, and has essentially the most current X11 support available for darwin. Many kudos to the upstream developer who has spearheaded that!
